presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankBuilder.java

```diff
@@ -32,40 +32,67 @@
 import static java.util.Objects.requireNonNull;
 
 /**
- * This class finds the top N rows defined by {@param comparator} for each group specified by {@param groupByHash}.
+ * This class finds the top N rows by rank value defined by {@param comparator} and {@param equalsAndHash} for each
+ * group specified by {@param groupByHash}.
  */
 public class GroupedTopNRankBuilder
         implements GroupedTopNBuilder
 {
     private static final long INSTANCE_SIZE = ClassLayout.parseClass(GroupedTopNRankBuilder.class).instanceSize();
 
     private final List<Type> sourceTypes;
-    private final boolean produceRowNumber;
+    private final boolean produceRanking;
     private final GroupByHash groupByHash;
     private final RowReferencePageManager pageManager = new RowReferencePageManager();
-    private final GroupedTopNRowNumberAccumulator groupedTopNRowNumberAccumulator;
+    private final GroupedTopNRankAccumulator groupedTopNRankAccumulator;
 
     public GroupedTopNRankBuilder(
             List<Type> sourceTypes,
             PageWithPositionComparator comparator,
+            PageWithPositionEqualsAndHash equalsAndHash,
             int topN,
-            boolean produceRowNumber,
+            boolean produceRanking,
             GroupByHash groupByHash)
     {
         this.sourceTypes = requireNonNull(sourceTypes, "sourceTypes is null");
         checkArgument(topN > 0, "topN must be > 0");
-        this.produceRowNumber = produceRowNumber;
+        this.produceRanking = produceRanking;
         this.groupByHash = requireNonNull(groupByHash, "groupByHash is null");
 
         requireNonNull(comparator, "comparator is null");
-        groupedTopNRowNumberAccumulator = new GroupedTopNRowNumberAccumulator(
-                (leftRowId, rightRowId) -> {
-                    Page leftPage = pageManager.getPage(leftRowId);
-                    int leftPosition = pageManager.getPosition(leftRowId);
-                    Page rightPage = pageManager.getPage(rightRowId);
-                    int rightPosition = pageManager.getPosition(rightRowId);
-                    return comparator.compareTo(leftPage, leftPosition, rightPage, rightPosition);
+        requireNonNull(equalsAndHash, "equalsAndHash is null");
+        groupedTopNRankAccumulator = new GroupedTopNRankAccumulator(
+                new RowComparisonStrategy()
+                {
+                    @Override
+                    public int compare(long leftRowId, long rightRowId)
+                    {
+                        Page leftPage = pageManager.getPage(leftRowId);
+                        int leftPosition = pageManager.getPosition(leftRowId);
+                        Page rightPage = pageManager.getPage(rightRowId);
+                        int rightPosition = pageManager.getPosition(rightRowId);
+                        return comparator.compareTo(leftPage, leftPosition, rightPage, rightPosition);
```

Comment:
this can be compiled (`io.prestosql.sql.gen.OrderingCompiler#internalCompilePageWithPositionComparator`), so maybe `equals` should default to `compare`?

Reply:
The equals and hash both need to be compiled actually too. Haven't prioritized that yet since for some reason it isn't the dominant bottleneck here.